### PR TITLE
Bybit :: fetchBalance update and openapi signing

### PR DESCRIPTION
**Pull request by carlosmiei**
- Updated to handle all market types (spot, swap/futures, usdc)
- Added openapi signing

Demo:

```
 node cli.js bybit fetchBalance '{"settle":"USDC"}' --verbose --swap
CCXT v1.81.71
bybit.fetchBalance ([object Object])
fetch Request:
 bybit POST https://api.bybit.com/option/usdc/openapi/private/v1/query-wallet-balance 
RequestHeaders:
RequestBody:
 {} 

handleRestResponse:
 bybit POST https://api.bybit.com/option/usdc/openapi/private/v1/query-wallet-balance 200 OK 
ResponseHeaders:
 {
  'Cache-Control': 'max-age=0, no-cache',
  Connection: 'keep-alive',
  'Content-Length': '277',
  'Content-Type': 'application/json; charset=utf-8',
  Date: 'Sun, 08 May 2022 09:48:16 GMT',
  Expires: 'Sun, 08 May 2022 09:48:16 GMT',
  Pragma: 'no-cache',
  Server: 'bat',
  Timenow: '1652003296065396774',
  Traceid: '37d85410a03ec30d'
} 
ResponseBody:
 {"result":{"walletBalance":"10.0000","accountMM":"0.0000","bonus":"0.0000","accountIM":"0.0000","totalSessionRPL":"0.0000","equity":"10.0000","totalRPL":"0.0000","marginBalance":"10.0000","availableBalance":"10.0000","totalSessionUPL":"0.0000"},"retCode":0,"retMsg":"Success."} 

2022-05-08T09:48:16.080Z iteration 0 passed in 210 ms

{
  info: {
    result: {
      walletBalance: '10.0000',
      accountMM: '0.0000',
      bonus: '0.0000',
      accountIM: '0.0000',
      totalSessionRPL: '0.0000',
      equity: '10.0000',
      totalRPL: '0.0000',
      marginBalance: '10.0000',
      availableBalance: '10.0000',
      totalSessionUPL: '0.0000'
    },
    retCode: '0',
    retMsg: 'Success.'
  },
  USDC: { free: 10, used: 0, total: 10 },
  free: { USDC: 10 },
  used: { USDC: 0 },
  total: { USDC: 10 }
}
```

### CODE REVIEW

File: js/bybit.js [hunk 1]
```diff
@@ -3327,24 +3435,48 @@ module.exports = class bybit extends Exchange {
...
-                const auth = this.rawencode (sortedQuery);
-                const signature = this.hmac (this.encode (auth), this.encode (this.secret));
-                if (method === 'POST') {
-                    body = this.json (this.extend (query, {
-                        'sign': signature,
-                    }));
+                const isOpenapi = url.indexOf ('openapi') >= 0;
+                const timestamp = this.milliseconds ();
+                if (isOpenapi) {
+                    const paramsLength = params.length;
```
Comment (kroitor) [inline js/bybit.js:3441]: The `params` arg is not an array, it's a dictionary.

File: js/bybit.js [hunk 2]
```diff
@@ -3327,24 +3435,40 @@ module.exports = class bybit extends Exchange {
...
-                const signature = this.hmac (this.encode (auth), this.encode (this.secret));
-                if (method === 'POST') {
-                    body = this.json (this.extend (query, {
-                        'sign': signature,
-                    }));
+                const isOpenapi = url.indexOf ('openapi') >= 0;
+                const timestamp = this.milliseconds ();
+                if (isOpenapi) {
+                    let query = '';
+                    if (Object.keys (params).length) {
```
Comment (carlosmiei) [inline js/bybit.js:3442]: This process is tricky, with your fix works on the PHP side but fails on Javascript (image attached)

From what I could test in PHP query needs to be empty '' and in JS needs to be an empty object {}.  That is the reason I was doing (Array.IsArray(...))

image

Comment (carlosmiei) [reply js/bybit.js:3442]: @kroitor Check my last commit, this is the only way I can make it work on the PHP side.

Comment (kroitor) [reply js/bybit.js:3442]: I'll try to add a shorter solution to it.

File: js/bybit.js [hunk 3]
```diff
@@ -3438,9 +3438,16 @@ module.exports = class bybit extends Exchange {
...
                 const timestamp = this.milliseconds ();
                 if (isOpenapi) {
-                    let query = '';
+                    let query = {};
                     if (Object.keys (params).length) {
-                        query = this.json (params);
+                        query = params;
+                    }
+                    // this is a PHP specific check
+                    const queryLength = query.length;
```
[the remark this answers is not in the document]
Comment (carlosmiei) [reply js/bybit.js:3446]: @kroitor it is working in Python too. In python `len(dict)` returns the number of keys a dict has, so I don't think it is a problem here
image

**Pull request merged**